### PR TITLE
Design of supporting ALB to expose service workbench website without domain

### PR DESCRIPTION
**Pull request by dengmingtong**
View link: 
https://github.com/dengmingtong/service-workbench-on-aws-cn/blob/mingtong_alb_rfc/rfcs/alb_expose_portal/index.md

Issue #, if available:

Description of changes:

Checklist:

- [ ] Have you successfully deployed to an AWS account with your changes?
- [ ] Have you successfully deployed to an AWS China account with your changes?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully tested with your changes locally?
- [ ] If new dependencies have been added, have they been pinned to specific versions?
- [ ] Is this change also required on the AWS Solution version?
- [ ] Have you updated openapi.yaml if you made updates to API definition (including add, delete or update parameter and request data schema)?
- [ ] If you had to run manual tests, have you considered automating those tests by adding them to [end-to-end tests](../main/end-to-end-tests/README.md)?
- [ ] If you are updating the changelog and vending out a new release, have you updated versionNumber and versionDate in [.defaults.yml](../main/config/settings/.defaults.yml)

AS review ticket id:

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: rfcs/alb_expose_portal/index.md [hunk 1]
```diff
@@ -0,0 +1,201 @@
+---
+feature name: alb_expose_portal
+start date: Nov 10th, 2022
+Summary: Use Application Load Balancer to expose Service Workbench portal without domain
+rfc pr: 
+related issue: https://github.com/awslabs/service-workbench-on-aws-cn/issues/117
+---
+
+# Summary
+This RFC proposes is to design a new architecture to make service workbench can be deployed and used in china regions without ICP domain.
```
Comment (zxkane) [inline rfcs/alb_expose_portal/index.md]: add scenario,

- deploy the entire solution inside VPC with secure network access

File: rfcs/alb_expose_portal/index.md [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+## High level design of service workbench
+
+![SWB_Architecture_Exsiting](./SWB_Architecture_Exsiting.png)
+
+According to this diagram, we can know there are two parts related with our new features:
+1. Front-end part: All front-end static files are stored in S3 bucket, and cloudfront distributes these static files to customers browser. Customers need to register cloudfront domain to the ICP domain to make cloudfront domain accessable in china regions.
+2. Deployment part: Build website statices files, create cloudfront, S3 bucket.
+
+## Front-end loading part
+In china regions, when users input the url of service workbench into browser, such as https://example-serviceworkbench.com, these requests are sent to cloudfront through domain name resolution, then forward request to S3 to get static files, such as html, css and js files.
```
Comment (zxkane) [inline rfcs/alb_expose_portal/index.md]: then forward request to S3 to get static files, such as html, css and js files.

-> add `while the cache is missing`

File: rfcs/alb_expose_portal/index.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+## Deployment part:
+The service workbench is based on serverless framework to deploy to AWS account.
+There are several serverless project to setup service workbench as below:
+1. infrastructure
+2. pre-deployment
+3. backend
+4. post-deployment
+5. ui
+![existing-deployment](./existing-deployment.drawio.png)
```
Comment (zxkane) [inline rfcs/alb_expose_portal/index.md:53]: might add a new line here for rendering version

File: rfcs/alb_expose_portal/index.md [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+## High level design.
+![new-front-end-processing.drawio.png](./new-front-end-processing.drawio.png)
+As above diagram shown.
+The new architecture will be compatible with existing functions.
+The Application Load Balancer(ALB) is a feature of Elastic Load Balancing that allows a developer to configure and route incoming end-user traffic to applications based in the AWS public cloud. We choose ALB to expose service workbench portal. Why do we choose application loader balancer, there are below reasons:
+1. ALB providers the domain, so customers do not need to apply ICP domain.
+2. Customers can set any listener port, such as 8000. so it is not necessary to apply ICP exception.
+3. We can set ALB VPC and security group to limit access in network layer.
+
+Lambda is a compute service that lets you run code without provisioning or managing servers. Lambda runs your code on a high-availability compute infrastructure and performs all of the administration of the compute resources, including server and operating system maintenance, capacity provisioning and automatic scaling, and logging. We choose lambda function with container to host front-end static files. The benefit are:
```
Comment (zxkane) [inline rfcs/alb_expose_portal/index.md:68]: `container image` might be better

https://docs.aws.amazon.com/lambda/latest/dg/images-create.html

File: rfcs/alb_expose_portal/index.md [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+If customers choose to enable ALB to expose servcie workbench website, the callflow will be like below:
+![new-runtime-processing.drawio.png](./new-runtime-processing.drawio.png)
+As above diagram shown:
+1. User access service workbench through Application Load Balancer.
+2. Application Load Balancer forward reqeust to Lambda function.
+3. Lambda starts a container, container starts a nginx server.
+4. Nginx return all static files back to browser through Application Load Balancer.
+5. After all website static files are loaded into browser, the website page should be rendered and displayed successfully. The following process will be the same with existing implementation.
+#### Limitation
+The maximum size of the response body that the Lambda function can send is 1 MB to ALB [Limitation}(https://docs.aws.amazon.com/elasticloadbalancing/latest/application/lambda-functions.html). So we need to make each static file must be less than 1MB. For this, we use **react-app-rewired** to achieve this goal.
```
Comment (zxkane) [inline rfcs/alb_expose_portal/index.md]: markdown syntax is wrong

**Review by zxkane — CHANGES_REQUESTED** [review]

File: rfcs/alb_expose_portal/index.md (hunk 5 above)
Comment (dengmingtong) [reply rfcs/alb_expose_portal/index.md]: Fixed

File: rfcs/alb_expose_portal/index.md (hunk 4 above)
Comment (dengmingtong) [reply rfcs/alb_expose_portal/index.md:68]: Fixed

File: rfcs/alb_expose_portal/index.md (hunk 2 above)
Comment (dengmingtong) [reply rfcs/alb_expose_portal/index.md]: add the sentence

File: rfcs/alb_expose_portal/index.md (hunk 1 above)
Comment (dengmingtong) [reply rfcs/alb_expose_portal/index.md]: Add new scenario

File: rfcs/alb_expose_portal/index.md (hunk 3 above)
Comment (dengmingtong) [reply rfcs/alb_expose_portal/index.md:53]: fixed

**Review by zxkane — APPROVED** [review]

**Pull request merged**